### ISSUE TITLE
[Wayland+HiDPI] Mouse cursor half as large as it should be

### ISSUE THREAD

**Issue by keynslug** [kovidgoyal/kitty#2647]
Hi! Thanks for the quality software.

Under GNOME Wayland session with 2x display scale factor I see that kitty renders much smaller (2x smaller I believe) mouse cursor, hovewer I'm sure that cursor _geometry_ itself is correctly scaled. This manifests in those troubles I have when I'm trying to select some word and missing constantly by a line, selecting a word line below instead.

Screenshots for comparison with GTK-based terminal:

![Screenshot from 2020-05-11 20-57-29](https://user-images.githubusercontent.com/1148366/81604461-93327d00-93d8-11ea-9d20-1330f3b36366.png)

![Screenshot from 2020-05-11 20-57-46](https://user-images.githubusercontent.com/1148366/81604466-9594d700-93d8-11ea-8ab8-e6c3c0fd3370.png)

```
 ~ gsettings get org.gnome.desktop.interface scaling-factor                                                                                                               
uint32 2
 ~ gsettings get org.gnome.desktop.interface cursor-size                                                                                                                  
24
 ~ gsettings get org.gnome.desktop.interface cursor-theme                                                                                                                 
'Adwaita'
```

```
 ~ kitty --debug-gl --debug-config                                                                                                                                        
kitty 0.17.4 created by Kovid Goyal
Linux kenfawks 5.6.8-arch1-1 #1 SMP PREEMPT Wed, 29 Apr 2020 16:22:56 +0000 x86_64
Arch Linux \r (\l)
LSB_VERSION=1.4
DISTRIB_ID=Arch
DISTRIB_RELEASE=rolling
DISTRIB_DESCRIPTION="Arch Linux"
Loaded config files: /home/keynslug/.config/kitty/kitty.conf
Running under: Wayland

Config options different from defaults:
background              Color(red=12, green=17, blue=21)
background_opacity      0.9
color0                  Color(red=46, green=52, blue=60)
color1                  Color(red=189, green=15, blue=47)
color10                 Color(red=73, green=233, blue=152)
color11                 Color(red=253, green=223, blue=110)
color12                 Color(red=42, green=139, blue=193)
color13                 Color(red=234, green=71, blue=39)
color14                 Color(red=160, green=182, blue=211)
color2                  Color(red=53, green=167, blue=112)
color3                  Color(red=251, green=148, blue=53)
color4                  Color(red=31, green=88, blue=114)
color5                  Color(red=189, green=37, blue=35)
color6                  Color(red=119, green=131, blue=151)
color7                  Color(red=255, green=255, blue=255)
color8                  Color(red=64, green=74, blue=85)
color9                  Color(red=189, green=15, blue=47)
cursor                  Color(red=108, green=108, blue=108)
editor                  /usr/bin/nano
font_size               9.0
foreground              Color(red=255, green=255, blue=255)
hide_window_decorations 1
initial_window_height   (800, 'px')
initial_window_width    (1200, 'px')
input_delay             5
kitty_mod               8
linux_display_server    wayland
remember_window_size    False
selection_background    Color(red=189, green=37, blue=35)selection_foreground    Color(red=12, green=17, blue=21)
shell                   /usr/bin/fish
tab_bar_min_tabs        1
tab_bar_style           powerline
touch_scroll_multiplier 15.0
Added shortcuts:
	super+x KeyAction(func='goto_layout', args=['stack'])
	super+backslash KeyAction(func='launch', args=['--type=window', '--cwd=current'])
	shift+super+minus KeyAction(func='change_font_size', args=(True, '-', 1.0))
	shift+super+0 KeyAction(func='tenth_window', args=())
	shift+super+1 KeyAction(func='first_window', args=())
	shift+super+2 KeyAction(func='second_window', args=())
	shift+super+3 KeyAction(func='third_window', args=())
	shift+super+4 KeyAction(func='fourth_window', args=())
	shift+super+5 KeyAction(func='fifth_window', args=())
	shift+super+6 KeyAction(func='sixth_window', args=())
	shift+super+7 KeyAction(func='seventh_window', args=())
	shift+super+8 KeyAction(func='eighth_window', args=())
	shift+super+9 KeyAction(func='ninth_window', args=())
	shift+super+equal KeyAction(func='change_font_size', args=(True, '+', 1.0))
	shift+super+c KeyAction(func='copy_to_clipboard', args=())
	shift+super+h KeyAction(func='show_scrollback', args=())
	shift+super+l KeyAction(func='next_layout', args=())
	shift+super+n KeyAction(func='new_os_window', args=())
	shift+super+v KeyAction(func='paste_from_clipboard', args=())
	shift+super+w KeyAction(func='close_tab', args=())
	shift+super+x KeyAction(func='goto_layout', args=['grid'])
	shift+super+grave_accent KeyAction(func='move_window_to_top', args=())
	shift+super+escape KeyAction(func='kitty_shell', args=['window'])
	shift+super+right KeyAction(func='next_tab', args=())
	shift+super+left KeyAction(func='previous_tab', args=())
	shift+super+down KeyAction(func='scroll_line_down', args=())
	shift+super+up KeyAction(func='scroll_line_up', args=())
	shift+super+page_up KeyAction(func='scroll_page_up', args=())
	shift+super+page_down KeyAction(func='scroll_page_down', args=())
Changed shortcuts:
	super+enter KeyAction(func='launch', args=['--type=tab', '--cwd=current'])
	super+delete KeyAction(func='combine', args=(KeyAction(func='clear_terminal', args=['scrollback', True]), KeyAction(func='send_text', args=['normal,application', b'\x0c'])))
	super+right KeyAction(func='next_window', args=())
	super+left KeyAction(func='previous_window', args=())
 ```

**Comment by kovidgoyal** [comment]
Under GNOME kitty will default to using XWayland since GNOME has no
support for serser side window decorations. Use linux_display_backend in
kitty.conf to force it to use wayland, or change the settings for cursor
size in xrdb

**Comment by keynslug** [comment]
> Use linux_display_backend in kitty.conf to force it to use wayland

I did. There's `linux_display_server    wayland` in changed settings.

**Comment by kovidgoyal** [comment]
I dont have GNOME to test with, but the cursor size is correct for me
under both weston and sway, with dpi set to 2

**Comment by keynslug** [comment]
Update for visibility.

Turns out cherry picking by hand those 3 commits from glfw upstream makes the issue go away.

* https://github.com/glfw/glfw/commit/321062833a6c452d805b5af6b8fbd646e31aeddc
* https://github.com/glfw/glfw/commit/15cad0223543745d13d4ce2de040d3b7defb3870
* https://github.com/glfw/glfw/commit/80fde12fdade85a39c5c1300d0a29200691fdc23

That said the patch working for me looks like this:
<details>

 <summary>cursor-2x-scale.patch</summary>

```
diff --git a/glfw/wl_init.c b/glfw/wl_init.c
index 49136bf0..0c0a757e 100644
--- a/glfw/wl_init.c
+++ b/glfw/wl_init.c
@@ -792,6 +792,8 @@ int _glfwPlatformInit(void)
                             "Wayland: Unable to load default cursor theme");
             return false;
         }
+        _glfw.wl.cursorThemeHiDPI =
+            wl_cursor_theme_load(cursorTheme, 2 * cursorSize, _glfw.wl.shm);
         _glfw.wl.cursorSurface =
             wl_compositor_create_surface(_glfw.wl.compositor);
     }
diff --git a/glfw/wl_platform.h b/glfw/wl_platform.h
index 4d17ac39..3d17fc3a 100644
--- a/glfw/wl_platform.h
+++ b/glfw/wl_platform.h
@@ -237,6 +237,7 @@ typedef struct _GLFWlibraryWayland
     int                         seatVersion;
 
     struct wl_cursor_theme*     cursorTheme;
+    struct wl_cursor_theme*     cursorThemeHiDPI;
     struct wl_surface*          cursorSurface;
     GLFWCursorShape             cursorPreviousShape;
     uint32_t                    pointerSerial;
@@ -299,6 +300,7 @@ typedef struct _GLFWmonitorWayland
 typedef struct _GLFWcursorWayland
 {
     struct wl_cursor*           cursor;
+    struct wl_cursor*           cursorHiDPI;
     struct wl_buffer*           buffer;
     int                         width, height;
     int                         xhot, yhot;
diff --git a/glfw/wl_window.c b/glfw/wl_window.c
index 166683c2..2f644b27 100644
--- a/glfw/wl_window.c
+++ b/glfw/wl_window.c
@@ -680,18 +680,24 @@ static bool createXdgSurface(_GLFWwindow* window)
 }
 
 static void
-setCursorImage(_GLFWcursorWayland* cursorWayland)
+setCursorImage(_GLFWwindow *window, _GLFWcursorWayland *cursorWayland)
 {
     struct wl_cursor_image* image;
     struct wl_buffer* buffer;
+    struct wl_cursor *wlCursor = cursorWayland->cursor;
     struct wl_surface* surface = _glfw.wl.cursorSurface;
+    int scale = 1;
 
-    if (!cursorWayland->cursor) {
+    if (!wlCursor) {
         buffer = cursorWayland->buffer;
         toggleTimer(&_glfw.wl.eventLoopData, _glfw.wl.cursorAnimationTimer, 0);
     } else
     {
-        image = cursorWayland->cursor->images[cursorWayland->currentImage];
+        if (window->wl.scale > 1 && cursorWayland->cursorHiDPI) {
+            wlCursor = cursorWayland->cursorHiDPI;
+            scale = 2;
+        }
+        image = wlCursor->images[cursorWayland->currentImage];
         buffer = wl_cursor_image_get_buffer(image);
         if (image->delay) {
             changeTimerInterval(&_glfw.wl.eventLoopData, _glfw.wl.cursorAnimationTimer, ms_to_monotonic_t(image->delay));
@@ -709,10 +715,15 @@ setCursorImage(_GLFWcursorWayland* cursorWayland)
         cursorWayland->yhot = image->hotspot_y;
     }
 
+    // _glfwInputError(GLFW_PLATFORM_ERROR,
+    //                 "Wayland: cursor scale=%d size=%d/%d hot=%d/%d",
+    //                 scale, cursorWayland->width, cursorWayland->height, cursorWayland->xhot, cursorWayland->yhot);
+
     wl_pointer_set_cursor(_glfw.wl.pointer, _glfw.wl.pointerSerial,
                           surface,
-                          cursorWayland->xhot,
-                          cursorWayland->yhot);
+                          cursorWayland->xhot / scale,
+                          cursorWayland->yhot / scale);
+    wl_surface_set_buffer_scale(surface, scale);
     wl_surface_attach(surface, buffer, 0, 0);
     wl_surface_damage(surface, 0, 0,
                       cursorWayland->width, cursorWayland->height);
@@ -728,7 +739,7 @@ incrementCursorImage(_GLFWwindow* window)
         {
             cursor->wl.currentImage += 1;
             cursor->wl.currentImage %= cursor->wl.cursor->image_count;
-            setCursorImage(&cursor->wl);
+            setCursorImage(window, &cursor->wl);
             toggleTimer(&_glfw.wl.eventLoopData, _glfw.wl.cursorAnimationTimer, cursor->wl.cursor->image_count > 1);
             return;
         }
@@ -849,6 +860,26 @@ struct wl_cursor* _glfwLoadCursor(GLFWCursorShape shape)
 #undef C
 }
 
+static char *translateCursorShape(int shape)
+{
+    switch (shape)
+    {
+    case GLFW_ARROW_CURSOR:
+        return "left_ptr";
+    case GLFW_IBEAM_CURSOR:
+        return "xterm";
+    case GLFW_CROSSHAIR_CURSOR:
+        return "crosshair";
+    case GLFW_HAND_CURSOR:
+        return "hand2";
+    case GLFW_HRESIZE_CURSOR:
+        return "sb_h_double_arrow";
+    case GLFW_VRESIZE_CURSOR:
+        return "sb_v_double_arrow";
+    }
+    return NULL;
+}
+
 //////////////////////////////////////////////////////////////////////////
 //////                       GLFW platform API                      //////
 //////////////////////////////////////////////////////////////////////////
@@ -1320,6 +1351,14 @@ int _glfwPlatformCreateStandardCursor(_GLFWcursor* cursor, GLFWCursorShape shape
     if (!standardCursor) return false;
     cursor->wl.cursor = standardCursor;
     cursor->wl.currentImage = 0;
+
+    if (_glfw.wl.cursorThemeHiDPI)
+    {
+        standardCursor = wl_cursor_theme_get_cursor(_glfw.wl.cursorThemeHiDPI,
+                                                     translateCursorShape(shape));
+        cursor->wl.cursorHiDPI = standardCursor;
+    }
+
     return true;
 }
 
@@ -1433,6 +1472,7 @@ static bool isPointerLocked(_GLFWwindow* window)
 void _glfwPlatformSetCursor(_GLFWwindow* window, _GLFWcursor* cursor)
 {
     struct wl_cursor* defaultCursor;
+    struct wl_cursor* defaultCursorHiDPI = NULL;
 
     if (!_glfw.wl.pointer)
         return;
@@ -1451,19 +1491,23 @@ void _glfwPlatformSetCursor(_GLFWwindow* window, _GLFWcursor* cursor)
     if (window->cursorMode == GLFW_CURSOR_NORMAL)
     {
         if (cursor)
-            setCursorImage(&cursor->wl);
+            setCursorImage(window, &cursor->wl);
         else
         {
             defaultCursor = _glfwLoadCursor(GLFW_ARROW_CURSOR);
             if (!defaultCursor) return;
+            if (_glfw.wl.cursorThemeHiDPI)
+                defaultCursorHiDPI =
+                    wl_cursor_theme_get_cursor(_glfw.wl.cursorThemeHiDPI,
+                                               "left_ptr");
             _GLFWcursorWayland cursorWayland = {
                 defaultCursor,
+                defaultCursorHiDPI,
                 NULL,
                 0, 0,
                 0, 0,
-                0
-            };
-            setCursorImage(&cursorWayland);
+                0};
+            setCursorImage(window, &cursorWayland);
         }
     }
     else if (window->cursorMode == GLFW_CURSOR_DISABLED)
```

</details>

I guess updating glfw to 3.3 would suffice as well, and sounds like a cleaner solution though way harder to achieve.

**Comment by kovidgoyal** [comment]
This patch looks pretty incomplete. It will work only with a scale factor
of 2. One really needs to maintain a growable list of cursor themes for
each seen scale factor. And rather than using translateCursorShape you
should use _glfwLoadCursor() which is much better than
translateCursorShape()

And there is the whole question of animated cursors.

**Comment by kovidgoyal** [comment]
Indeed, looking at the code, all you need to do is set the XCURSOR_SIZE environament variable to the size you want.

**Comment by kovidgoyal** [comment]
And the proper solution for this requires a new wayland protocol, perhaps: https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/21

**Comment by keynslug** [comment]
> This patch looks pretty incomplete. It will work only with a scale factor
> of 2. One really needs to maintain a growable list of cursor themes for
> each seen scale factor.

You're right. But this is how it is on the glfw master, I even seen this problem mentioned somewhere in comments.

> https://github.com/glfw/glfw/blob/7dbdd2e6a5f01d2a4b377a197618948617517b0e/src/wl_window.c#L657-L662

I probably should have mentioned that I did my best to not depart far from glfw upstream with this patch.

> And rather than using translateCursorShape you
> should use _glfwLoadCursor() which is much better than
> translateCursorShape()

I would like to. But it's unaware of available themes. In contrast here's what it looks like on the master.

> https://github.com/glfw/glfw/blob/7dbdd2e6a5f01d2a4b377a197618948617517b0e/src/wl_window.c#L1212

> Indeed, looking at the code, all you need to do is set the XCURSOR_SIZE environament variable to the size you want.

Actually this is what I attempted initially, but the issue with cursor geometry being twice as large as it should be persisted.

> And the proper solution for this requires a new wayland protocol, perhaps: https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/21

Oh great, fingers crossed.

**Comment by kovidgoyal** [comment]
On Wed, May 13, 2020 at 02:09:36AM -0700, Andrew Mayorov wrote:
> > This patch looks pretty incomplete. It will work only with a scale factor
> > of 2. One really needs to maintain a growable list of cursor themes for
> > each seen scale factor.
> 
> You're right. But this is how it is on the glfw master, I even seen this problem mentioned somewhere in comments.
> 
> > https://github.com/glfw/glfw/blob/7dbdd2e6a5f01d2a4b377a197618948617517b0e/src/wl_window.c#L657-L662
> 
> I probably should have mentioned that I did my best to not depart far from glfw upstream with this patch.

I prefer solving it correctly once and for all. I have diverged a fair
bit from upstream glfw already.

> 
> > And rather than using translateCursorShape you
> > should use _glfwLoadCursor() which is much better than
> > translateCursorShape()
> 
> I would like to. But it's unaware of available themes. In contrast here's what it looks like on the master.

No reason it cant be made aware of them.

> 
> > https://github.com/glfw/glfw/blob/7dbdd2e6a5f01d2a4b377a197618948617517b0e/src/wl_window.c#L1212
> 
> > Indeed, looking at the code, all you need to do is set the XCURSOR_SIZE environament variable to the size you want.
> 
> Actually this is what I attempted initially, but the issue with cursor geometry being twice as large as it should be persisted.

Interesting is this because of the scale of the surface used for the
cursor?

**Comment by keynslug** [comment]
> I prefer solving it correctly once and for all. I have diverged a fair bit from upstream glfw already.

Fair enough. I'd very much like to make a proper PR in my spare time then if you're open to accepting one. Though can't promise any timeframe honestly.

> Interesting is this because of the scale of the surface used for the cursor?

Probably, but I'm not sure. Another observation: `wl_surface_set_buffer_scale(surface, scale)` (assuming scale = 2) doesn't make any effect on my system. Conversely, scaling cursor hotspot down by scale factor alleviates the issue.

**Comment by kovidgoyal** [comment]
On Wed, May 13, 2020 at 03:52:28AM -0700, Andrew Mayorov wrote:
> > I prefer solving it correctly once and for all. I have diverged a fair bit from upstream glfw already.
> 
> Fair enough. I'd very much like to make a proper PR in my spare time then if you're open to accepting one. Though can't promise any timeframe honestly.

Sure, I'll be happy to accept a PR for this.

> 
> > Interesting is this because of the scale of the surface used for the cursor?
> 
> Probably, but I'm not sure. Another observation: `wl_surface_set_buffer_scale(surface, scale)` (assuming scale = 2) doesn't make any effect on my system. Conversely, scaling cursor hotspot down by scale factor alleviates the issue.

Well since you can replicate it and I cant, probably better if you make
the PR.

**Comment by TheDaemoness** [comment]
I should be able to submit a PR that would properly fix this/#2669 within a week, likely less than that.

@keynslug I'm not sure how much you've already done, but if you've already started, what are your preferences regarding collaboration on this issue?

**Comment by keynslug** [comment]
@TheDaemoness Hadn't time to make a progress yet, sorry. Nice to hear that you had, I'll take a look at your PR to see if can be of any help.
